make ChangePasswordController and UI Code respect Message and Hint fr…

blizzz opened the pull request:
…om HintException

Needed for #1715 

Please review @GitHubUser4234 @nickvergessen @ChristophWurst

blizzz committed:
make ChangePasswordController and UI Code respect Message and Hint from HintException

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz commented:
@MorrisJobke could you check drone? it tells me nothing and does not let me log in.

blizzz committed:
adjust test

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

michag86 commented on settings/js/users/users.js:
```diff
@@ -664,7 +664,11 @@ $(document).ready(function () {
+									if(!_.isEmpty(result.data.hint)) {
+										message += "<br/>" + _.escape(result.data.hint);
+									}
+									OC.Notification.showTemporary(t('admin', message), {isHTML: true});
```
Maybe it's better tu use `t` in line 667 and 669?

blizzz replied:
@michag86 this does not help. To make sure that a string is translated, it must be provided as plain text to the method. That way it can be parsed and fed into the translation tool. 

Basically I did not change the behavior much. If translation worked before, it is pure luck. This should happen on the server side code where the string is passed (if applicable). Best thing to do here is actually remove the transaltion of received strings.

blizzz replied:
Removed the translation call.

blizzz committed:
Don't attempt to translate variables

* stings should be translated (if applicable) where they happen
* ChangePasswordController returns all its fixed strings already translated
* local user backend does not throw HintException at all

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

rullzer commented:
Makes sense. And seems to do it job :+1:

MorrisJobke commented:
- [ ] Changing the password on the personal page causes this: (it shows the message and the hint(localized))

<img width="1032" alt="bildschirmfoto 2016-10-20 um 12 46 41" src="https://cloud.githubusercontent.com/assets/245432/19556857/587539fc-96c3-11e6-892f-93481790fbbf.png">

MorrisJobke commented:
Same for the admin user management (when changing the password of an existing user):

![bildschirmfoto 2016-10-20 um 12 48 12](https://cloud.githubusercontent.com/assets/245432/19556885/816ca016-96c3-11e6-85ae-d69d798ed0b2.png)

nickvergessen commented on settings/js/personal.js:
```diff
@@ -231,6 +231,9 @@ $(document).ready(function () {
 				} else {
 					if (typeof(data.data) !== "undefined") {
+						if(!_.isEmpty(data.data.hint)) {
+							data.data.message += "\n" + data.data.hint;
```
In normal cases hint is the translated version of message. That is at least the case for all HintExceptions I have seen so far.

blizzz commented:
@MorrisJobke thx.

Wow. After acking through the HintException usages it seems that half of the time the ``hint``parameter is being used as a translated message. Imho that does not make much sense, since this is supposed to be user facing anyway. Also, my perception is that hint should be a suggestion for an action, instead of the reason of the error.

However since it is vastly used in this manner i am hesitating to change this all, and this would include other repos… If this is expected to land into logExpection translated stuff does not make sense again either. Perhaps the best way is to extend (or add another Exception inheriting from it for backwards compatibility) for another parameter, translated message or so :see_no_evil: :hear_no_evil: :speak_no_evil: 

This amazingly extend the PR, resulting in more work than I anticipated. Perhaps I postpone it for a little bit later and we need to go with a workaround solution in #1715 first.

MorrisJobke commented:
@blizzz The hint is the reason that could be shown to the user and is therefore translated. The hint never lands in the logs. The message of an hint exception is untranslated and will never get to the user because it could contain sensitive data. So the usage of the HintException holding the translated string is correct.

cc @LukasReschke

blizzz commented:
@morrisjobke then we ought to rename the parameter and add some doc. This is confusing and misleading.

GitHubUser4234 commented:
Yeah, at least adding some comments to the constructor / getHint() method would give a "hint" on how to use them =)  

Alright, for the #1715, I'll adjust the code accordingly, thanks @blizzz though 👍

rullzer commented:
So what are we going to do here? @blizzz @MorrisJobke ?

MorrisJobke commented:
> So what are we going to do here? @blizzz @MorrisJobke ?

At least resolve the conflicts :(

MorrisJobke commented:
> Needed for #1715

This was already merged in #2286 - how to proceed here?

MorrisJobke commented:
I fixed the conflicts and rebased on master ;)

blizzz commented:
> I fixed the conflicts and rebased on master ;)

That's kind, but apart from PHP Doc adjustments there is nothing else to do (cf. https://github.com/nextcloud/server/pull/1787#issuecomment-255324534) I come up with a new PR instead.
